MatrixSDK/VoIP/MXCallManager.m

```diff
@@ -1267,6 +1380,199 @@ - (void)placeCallAgainst:(NSString *)phoneNumber
+                                              @"virtual_mxid": userId
+                                          }
+                                         success:^(MXThirdPartyUsersResponse *thirdpartyUsersResponse) {
+        
+        MXThirdPartyUserInstance * user = [thirdpartyUsersResponse.users firstObject];
+        
+        NSLog(@"Succeeded to look up the native user: %@", user.userId);
+        
+        if (user)
+        {
+            if (success)
+            {
+                success(user);
+            }
+        }
+        else
+        {
+            if (failure)
+            {
+                failure(nil);
+            }
+        }
+    } failure:^(NSError *error) {
+        NSLog(@"Failed to look up the native user with error: %@", error);
+        if (failure)
+        {
+            failure(error);
+        }
+    }];
+}
+
+/// Tries to find a direct & callable room with the given virtual user. If not such a room found, tries to create it as a virtual room and then waits for the other party to join.
+/// @param userId The virtual user id to check.
+/// @param timeout The timeout for the invited user to join the room, in case of the room is newly created (in seconds).
+/// @param completion Completion block.
+- (void)directCallableRoomWithVirtualUser:(NSString * _Nonnull)userId
+                             nativeRoomId:(NSString * _Nonnull)nativeRoomId
+                                  timeout:(NSTimeInterval)timeout
+                               completion:(void (^_Nonnull)(MXRoom* _Nullable room, NSError * _Nullable error))completion
+{
+    MXRoom *room = [self.mxSession directJoinedRoomWithUserId:userId];
+    if (room)
+    {
+        [room state:^(MXRoomState *roomState) {
+            MXMembership membership = [roomState.members memberWithUserId:userId].membership;
+            
+            if (membership == MXMembershipJoin)
+            {
+                //  set account data on the room, if required
+                //  room may be created earlier for a different native room which was left. So check the native room id.
+                if (![room.accountData.nativeRoomId isEqualToString:nativeRoomId])
+                {
+                    MXWeakify(room);
+                    [room setAccountData:@{
+                        kRoomNativeRoomIdJSONKey: nativeRoomId
+                    } forType:kRoomIsVirtualJSONKey success:^{
+                        MXStrongifyAndReturnIfNil(room);
+                        [self.mxSession setVirtualRoom:room.roomId forNativeRoom:nativeRoomId];
+                    } failure:^(NSError *error) {
+                        completion(nil, error);
+                    }];
+                }
+                
+                //  other party already joined, return the room
+                completion(room, nil);
+            }
+            else if (membership == MXMembershipInvite)
+            {
+                //  Wait for other party to join before returning the room
```

Comment:
I'm not sure this bit's actually necessary - the other party should still get the call once they join.

Reply:
That was something we did for native rooms and actually copied here. I can remove if not necessary.